src/coins.cpp

```diff
@@ -77,8 +75,26 @@ void CCoinsViewCache::AddCoin(const COutPoint &outpoint, Coin&& coin, bool possi
     }
     if (!possible_overwrite) {
         if (!it->second.coin.IsSpent()) {
-            throw std::logic_error("Adding new coin that replaces non-pruned entry");
+            throw std::logic_error("Attempted to overwrite an unspent coin");
         }
+        // If the coin exists in this cache, is spent and is DIRTY, then
+        // its spentness hasn't been flushed to the parent cache. We're
+        // re-adding the coin to this cache now but we can't mark it as FRESH.
+        // If we mark it FRESH and then spend it before the cache is flushed
+        // we would remove it from this cache and would never flush spentness
+        // to the parent cache.
+        //
+        // Re-adding a spent coin can happen in the case of a re-org (the coin
+        // is 'spent' when the block containing it is disconnected and then
+        // re-added if it also appears in the newly connected blocks). It can
+        // also happen in the case of duplicate txids. In practice, BIP30 and
+        // BIP34 ensure that the only duplicate transactions before block
+        // 1,983,702 are the two pairs of coinbase txs at heights 91812/91842
```

Comment:
Block 1,983,702 -- about 26 years from now? Is that a timestamp rolling over or something?
I thought BIP34 (block height in coinbase) will prevent this from ever happening since it was enforced:

> Block number 227,835 (timestamp 2013-03-24 15:49:13 GMT) was the last version 1 block.

Reply:
Take a look at the comments here: https://github.com/bitcoin/bitcoin/blob/a421e0a22f1230abd69b4661a019bed39b72205f/src/validation.cpp#L1989

Reply:
Oh THAT large comment in `connectBlock()` thanks ;-)